### PR TITLE
Add `set_XY` and `set_data` to `Quiver`

### PR DESCRIPTION
**Pull request by ianhi**
## PR Summary
Will close https://github.com/matplotlib/matplotlib/issues/22402 and close https://github.com/matplotlib/matplotlib/issues/11790

still a bit of tidying up to do, and should probably pair with a similar implementation for Barb.

Opening early to get feedback I think I did all the shape setting correctly, but not entirely clear whats happening with all the masking. Also this will enable modifying the number of points and there was originally some concern about that.

Example script:

```python
import matplotlib.pyplot as plt
import numpy as np

X, Y = np.random.rand(10, 2).T
U, V = np.random.rand(10, 2).T

fig, ax = plt.subplots()
Q = ax.quiver(X, Y, U, V)

plt.show(block=False)
plt.pause(2)

X, Y, U, V, C = np.random.rand(11, 5).T
Q.set_data(X, Y, U, V, C)
fig.canvas.draw()
plt.show(block=True)
```
## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by jklymak** [comment]
The API seems fine, though I wonder about providing the conveniences, versus just `set_data`.  Why does your example need a fig.canvas.draw?  (= fig.draw_without_rendering)

**Comment by ianhi** [comment]
> The API seems fine, though I wonder about providing the conveniences, versus just set_data.

I put in both since `set_UVC` already exists although i suppose instead that could be deprecated?


> fig.canvas.draw? (= fig.draw_without_rendering)

I'm a bit confused. the example changes after some time so since it's displayed already you need to call draw to update it

**Comment by jklymak** [comment]
I dont think of fig.canvas.draw as user-facing.

**Comment by ianhi** [comment]
`fig.canvas.draw` (or draw_idle) is crucial when doing anything interactive with `set_data` etc. For example see the slider examples: https://matplotlib.org/stable/gallery/widgets/slider_demo.html or https://matplotlib.org/stable/gallery/animation/strip_chart.html?highlight=codex%20fig%20canvas%20draw  here it's necessary to update the figure because the second show doesn't force a draw.

File: lib/matplotlib/quiver.py [hunk 1]
```diff
@@ -451,11 +461,10 @@ class Quiver(mcollections.PolyCollection):
     """
     Specialized PolyCollection for arrows.
 
-    The only API method is set_UVC(), which can be used
-    to change the size, orientation, and color of the
-    arrows; their locations are fixed when the class is
-    instantiated.  Possibly this method will be useful
-    in animations.
+    The only API methods are ``set_UVC()``, ``set_XY``,
```
Comment (oscargus) [inline lib/matplotlib/quiver.py:464]: Remove "only"? (Does three count as only?)

Single back-ticks will link, right? I guess one should be consistent with either using `()` for all or none of the methods.

**Comment by oscargus** [comment]
Couldn't comment in the source, but `Barbs` has this in the docs: 
```
The only API method is :meth:`set_UVC`, which can be used to
```
which no longer is correct.

**Comment by oscargus** [comment]
I think this would be nice to get into 3.7. Would you have time to update it?

I guess one can use the docs of the original Quiver class to explain the U, V, and C arguments in `set_data`.

**Comment by ianhi** [comment]
> I think this would be nice to get into 3.7. Would you have time to update it?

maybe - when is the deadline there? jan 1?

Also dependent on figuring out what I meant by: 
> still a bit of tidying up to do

also

>  and should probably pair with a similar implementation for Barb.

probably won't be able to do that

**Comment by ianhi** [comment]
> > and should probably pair with a similar implementation for Barb.
> probably won't be able to do that

or plot twist! did I already do this???

File: lib/matplotlib/quiver.py [hunk 2]
```diff
@@ -567,17 +578,58 @@ def draw(self, renderer):
...
+        """
+        Update the locations and/or rotation and color of the arrows.
+
+        Parameters
+        ----------
+        X, Y : arraylike of float
+            The arrow locations, any shape is valid so long
+            as X and Y have the same size.
+        U, V : ???
+        C : ???
```
Comment (ianhi) [inline lib/matplotlib/quiver.py:605]: note to self - this is part of the "cleaning up"

**Comment by jklymak** [comment]
@ianhi I'll move to Draft until you have time to come back to this...

**Comment by QuLogic** [comment]
This was replaced by #26410.

**Pull request closed without merging**